### PR TITLE
Keras file format updates

### PR DESCRIPTION
**Pull request by rundxdi**
## Fixes

Addresses file format changes associated with updates to tensorflow and keras mentioned in issue #1369 

## Summary/Motivation:

Tensorflow and keras version updates changed how they save/load files.  This PR updates the keras_surrogate.py file to accommodate those changes.

## Changes proposed in this PR:
- Update keras_surrogate.py to save new .keras file format
- Update test_keras_surrogate.py with new versions of saved models and accounting for new save/load methods

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: idaes/core/surrogate/keras_surrogate.py
Comment (andrewlee94) [inline idaes/core/surrogate/keras_surrogate.py:266]: Should the file name be hard-coded here, or should we make this an input from the user (either optional or required).

**Comment by bpaul4** [comment]
@rundxdi thank you for opening this. Once this is ready to test, it would be good to undo the temporary fix implemented in https://github.com/IDAES/idaes-pse/pull/1373 to ensure the latest version of TensorFlow is used in the tests.

File: idaes/core/surrogate/keras_surrogate.py
Comment (bpaul4) [reply idaes/core/surrogate/keras_surrogate.py:266]: I support letting users pass a string for the file name, so something like

```suggestion
        self._keras_model.save(os.path.join(keras_folder_name, keras_file_name + ".keras"))
```

where `keras_file_name` is an input to the save method. I don't know if we still need to specify the folder name, or if a single string for the entire path is sufficient.

**Comment by ksbeattie** [comment]
@rundxdi any progress on this?

**Comment by rundxdi** [comment]
> @rundxdi any progress on this?

Nothing this week -- still trying to figure out plotting tests.

File: idaes/core/surrogate/keras_surrogate.py
Comment (rundxdi) [reply idaes/core/surrogate/keras_surrogate.py:266]: The folder name is used to save/load some model weights and some idaes specific information separate from the filename.  But I think the file name should be rename-able as you say -- making that change now

**Comment by rundxdi** [comment]
I have two remaining issues:

1. Python 3.8 doesn't like test_keras_surrogate.py.  Tests all produce failure statuses looking like:

`Exception encountered: Unrecognized keyword arguments: ['batch_shape']`

These tests all pass on Python 3.9+.

2. The test_sofc_surrogates.py tests rely on a saved Keras surrogate that cannot be loaded with Keras v3+ (and thus cannot be loaded with Tensorflow 2.16.1).  I've found no obvious way to generate an equivalent surrogate in Keras v3 equivalent surrogate; their suggested workaround doesn't work.

**Comment by lbianchi-lbl** [comment]
>     1. Python 3.8 doesn't like test_keras_surrogate.py.  Tests all produce failure statuses looking like:
> 
> `Exception encountered: Unrecognized keyword arguments: ['batch_shape']`

> 
> These tests all pass on Python 3.9+.

My guess is that this is due to NumPy (and consequently I imagine most of the downstream projects depending on it) having dropped support for Python 3.8 a while back. So, when installing on Python 3.8, an older version of the package (the latest compatible) gets installed, which I assume doesn't support the same arguments as the current version.

We're considering removing support for (i.e. stopping testing with) Python 3.8. In the meantime, you should be able to use `pytest.mark.skipif()` to skip the failing tests if they're being run: https://docs.pytest.org/en/latest/how-to/skipping.html#id1


>     2. The test_sofc_surrogates.py tests rely on a saved Keras surrogate that cannot be loaded with Keras v3+ (and thus cannot be loaded with Tensorflow 2.16.1).  I've found no obvious way to generate an equivalent surrogate in Keras v3 equivalent surrogate; their suggested workaround doesn't work.

With the disclaimer that I don't have any significant hands-on experience using Keras, I'm at least aware of there being limited or no support for cross-version compatibility (i.e. a model serialized with Keras version X won't work when you try to deserialize it using version Y). I'm not sure how much the new (current) serialization format changes things. If this is the case, I guess the solution would be to regenerate the file used in the tests using the current version. I'm not sure if this addresses the original issue, though.

**Comment by andrewlee94** [comment]
As mentioned in the dev call, regarding the test failure in the power generation models you should:

1. Mark the failing tests with `pytest.mark.xfail`
2. In the failing code, add a deprecation warning indicating the tests for this code have started failing and the code will be removed no earlier than August if it is not fixed.
3. Open an issue to fix this and assign it to the code owner(s), and put this on the August release board as a High priority so we track it.

**Comment by ksbeattie** [comment]
@rundxdi we expect to cut the May release next week, so if this PR is to be included, it needs to be merged very soon.

**Comment by ksbeattie** [comment]
@rundxdi, this missed the May release, not on the Aug release board.

**Review by bpaul4 — COMMENTED** [review]
Thank you @rundxdi for investigating and implementing the fix. The changes to the Keras/OMLT calls and file writing look good.

I have a couple questions. First, the models previously import from a folder containing `saved_model.pb`, `keras_metadata.pb`, and `idaes_info.json` files and a `variables/` folder containing `variables.index` and `variables.data-00000-of-00001` files. Are these still necessary, or can we remove them in favor of the `.keras` files? I've seen in my research into Keras 3 that it now doesn't support SavedModel files and includes a read-only TFSM format instead.

Second, do we want to look into the Python 3.8 failures? Since it isn't occurring in the newer Python environments, it may just be a compatibility issue that's best skipping in that one environment.

**Comment by bpaul4** [comment]
@rundxdi when you have a moment, could you please provide a status update? What is required to finalize this?

**Comment by rundxdi** [comment]
> @rundxdi when you have a moment, could you please provide a status update? What is required to finalize this?

Hi @bpaul4 -- we can safely remove any of the non `.keras` files.  The Python 3.8 tests should be skipped at this point.  

It should be ready to finalize/review for finalization.

**Review by bpaul4 — APPROVED** [review]
I think this looks good now, thanks @rundxdi!

**Comment by lbianchi-lbl** [comment]
Before merging this, we should remove Python 3.8 from the CI and see if the tests pass without needing to mark them as `xfail`ing.

**Comment by bpaul4** [comment]
@andrewlee94 it looks like all of your prior review comments have been addressed.

I removed the xfail markers from the keras surrogate tests and everything looks good.

File: idaes/models_extra/power_generation/properties/tests/test_sofc_surrogates.py
Comment (andrewlee94) [inline idaes/models_extra/power_generation/properties/tests/test_sofc_surrogates.py:56]: Could you add an inline comment to explain why this is `xfailed`. I know this is mentioned in the deprecation warning in the code, but having comments here will help reinforce that this issue needs to be fixed.

**Review by andrewlee94 — COMMENTED** [review]
@bpaul4 I was waiting for the Python 3.8 issue to be merged to make sure the tests passed. I have just one minor request.

File: idaes/models_extra/power_generation/properties/tests/test_sofc_surrogates.py
Comment (bpaul4) [reply idaes/models_extra/power_generation/properties/tests/test_sofc_surrogates.py:56]: I'll add the comments now. @AlexNoring is aware of the SOFC surrogate issue and will continue the discussion here: https://github.com/IDAES/idaes-pse/issues/1431

**Review by andrewlee94 — APPROVED** [review]

**Comment by bpaul4** [comment]
@andrewlee94 how would you suggest handling the failing example integration tests? The failures are resolved as part of https://github.com/IDAES/examples/pull/128.

**Comment by andrewlee94** [comment]
@bpaul4 I would say we need to get that examples PR merged first.

File: idaes/core/surrogate/keras_surrogate.py
Comment (lbianchi-lbl) [inline idaes/core/surrogate/keras_surrogate.py:268]: Suggestion: since Keras seems to support `pathlib.Path` objects now, this could be rewritten to be slightly more concise/readable (requires replacing `import os.path` with `from pathlib import Path` at the top of the file):

```suggestion
        self._keras_model.save(
            Path(keras_folder_name, keras_model_name).with_suffix(".keras")
        )
```

File: idaes/core/surrogate/keras_surrogate.py
Comment (lbianchi-lbl) [inline idaes/core/surrogate/keras_surrogate.py:302]: See previous comments for using `pathlib.Path` instead of `os.path`.

File: idaes/core/surrogate/keras_surrogate.py
Comment (lbianchi-lbl) [inline idaes/core/surrogate/keras_surrogate.py:334]: See comments above for using `pathlib.Path` instead of `os.path`.

**Review by lbianchi-lbl — COMMENTED** [review]
Mostly minor comments, questions, and suggestions.

File: setup.py
Comment (bpaul4) [reply setup.py:96]: I believe that would be appropriate. although removing the tag will automatically search for the latest version. Pinning a minimum version would be a good safety net.

**Comment by lbianchi-lbl** [comment]
There might be another test that needs to be skipped/xfailed as it's currently causing a failure in the "user install" integration checks: 
image

**Comment by andrewlee94** [comment]
@lbianchi-lbl I think we should look into that one a bit more first, and fix it if we can. Being that is in the core code, we should at least know why it fails.

**Comment by lbianchi-lbl** [comment]
> @lbianchi-lbl I think we should look into that one a bit more first, and fix it if we can. Being that is in the core code, we should at least know why it fails.

The fact that it's only failing in the user-mode check, plus the exception raised, seem to point towards the file not being included in the non-developer installation:

image

**Pull request merged**
